Adding specs for new _list/indices and _list/shards APIs

gargharsh3134 opened the pull request:
### Description
Adding specs for ```_list/indices``` and ```_list/shards``` APIs, new paginated alternates of ```_cat/indices``` and ```_cat/shards```. Both the new list APIs will continue to support all the params supported by existing cat API counterparts in addition to three new parameters which are specific to pagination (namely ```next_token```, ```size``` and ```order```).

Related Opensearch changes: https://github.com/opensearch-project/OpenSearch/issues/14258 & https://github.com/opensearch-project/OpenSearch/issues/14257


### Issues Resolved
_List any issues this PR will resolve, e.g. Closes [...]._ 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

gargharsh3134 commented:
@dblock Can you please take a look again? Added tests taking inspiration from existing _cat APIs. Verified the tests by running them against Amazon Opensearch domain.

gargharsh3134 commented:
The staging 2.18 distribution seems to be missing the _list APIs, and as a result Spec test in CI are failing.

dblock commented on spec/namespaces/list.yaml:
```diff
@@ -0,0 +1,485 @@
+      operationId: list.indices.0
+      x-operation-group: list.indices
+      x-version-added: '2.18'
+      description: 'Returns information about indices in a paginated fashion: number of primaries and replicas, document counts, disk size, ...'
```
Let's make it a full sentence.

```suggestion
      description: 'Returns paginated information about indices including number of primaries and replicas, document counts, disk size, etc.'
```

gargharsh3134 replied:
Updated in the revision.

dblock commented on spec/namespaces/list.yaml:
```diff
@@ -0,0 +1,485 @@
+      operationId: list.indices.1
+      x-operation-group: list.indices
+      x-version-added: '2.18'
+      description: 'Returns information about indices in a paginated fashion: number of primaries and replicas, document counts, disk size, ...'
```
Same as above.

gargharsh3134 replied:
Updated in the revision.

dblock commented on spec/namespaces/list.yaml:
```diff
@@ -0,0 +1,485 @@
+      operationId: list.shards.0
+      x-operation-group: list.shards
+      x-version-added: '2.18'
+      description: Provides a detailed view of shard allocation on nodes in a paginated fashion.
```
```suggestion
      description: Returns paginated details of shard allocation on nodes.
```

gargharsh3134 replied:
Updated in the revision.

dblock commented on spec/namespaces/list.yaml:
```diff
@@ -0,0 +1,485 @@
+      operationId: list.shards.1
+      x-operation-group: list.shards
+      x-version-added: '2.18'
+      description: Provides a detailed view of shard allocation on nodes in a paginated fashion.
```
Same as above.

gargharsh3134 replied:
Addressed in the revision.

dblock commented on spec/namespaces/list.yaml:
```diff
@@ -0,0 +1,485 @@
+    list.indices::query.sort:
+      name: sort
+      in: query
+      description: Defines order in which indices will be displayed. If "desc", most recently created indices would be displayed first.
```
I think we quote these as `desc`, but double check.

gargharsh3134 replied:
Thanks for pointing out. Replaced "desc" with `desc` as is being used at other places.

dblock commented on spec/namespaces/list.yaml:
```diff
@@ -0,0 +1,485 @@
+    list.shards::query.sort:
+      name: sort
+      in: query
+      description: Defines order in which shards will be displayed. If "desc", most recently created shards would be displayed first.
```
Same as above.

gargharsh3134 replied:
Addressed in the revision.

dblock commented on tests/default/list/indices.yaml:
```diff
@@ -0,0 +1,151 @@
+      status: 200
+      content_type: application/cbor
+  - synopsis: List information related to indexes in different formats (format=smile).
+    version: '>= 2.18'
```
Since this whole thing is `>= 2.18`, move that to the top of this file and remove from individual tests.

gargharsh3134 replied:
Addressed in the revision.

dblock commented on tests/default/list/shards.yaml:
```diff
@@ -0,0 +1,51 @@
+      h:
+        - '*'
+  - synopsis: List the state of all primary and replica shards and how they are distributed for an index.
+    version: '>= 2.18'
```
Move up.

gargharsh3134 replied:
Addressed in the revision.

dblock reviewed (changes requested):
This is looking good!

The next version (2.18 and 3.0) distribution uses a specific SHA from https://hub.docker.com/r/opensearchstaging/opensearch/tags to remain stable. Update both to the latest ones that have the `_list` API, and make sure tests pass.

gargharsh3134 commented:
@dblock Thanks for reviewing. I have addressed previous set of comments, please do take a look again whenever you get a chance. Also, sorry for the delay in between the revisions, got occupied with other urgent items.

On the tests, i have updated the staging distributions for 2.18 and 3.0 and the newly added list tests are passing. Not sure why some of the existing ones have started to fail. Trying to root cause.

dblock commented on spec/namespaces/list.yaml:
```diff
@@ -0,0 +1,485 @@
+      operationId: list.indices.0
+      x-operation-group: list.indices
+      x-version-added: '2.18'
+      description: 'Returns paginated information about indices including number of primaries and replicas, document counts, disk size, etc.'
```
```suggestion
      description: Returns paginated information about indices including number of primaries and replicas, document counts, disk size, etc.
```

gargharsh3134 replied:
Updated in the revision.

dblock commented on spec/namespaces/list.yaml:
```diff
@@ -0,0 +1,485 @@
+      operationId: list.indices.1
+      x-operation-group: list.indices
+      x-version-added: '2.18'
+      description: 'Returns paginated information about indices including number of primaries and replicas, document counts, disk size, etc.'
```
```suggestion
      description: Returns paginated information about indices including number of primaries and replicas, document counts, disk size, etc.
```

gargharsh3134 replied:
Updated in the revision.

dblock reviewed (changes requested):
The 3.0 build gets generated even if not all plugins build with it, so looks like the update SHA doesn't have all the plugins? Try another more recent SHA until the tests all pass, possibly hunt down why the build is failing for the plugins that are failing (geospatial looks like).

gargharsh3134 commented:
> The 3.0 build gets generated even if not all plugins build with it, so looks like the update SHA doesn't have all the plugins? Try another more recent SHA until the tests all pass, possibly hunt down why the build is failing for the plugins that are failing (geospatial looks like).

Got it, I had sorted the distributions via the newest and used the latest one already. Can this staging build be triggered manually to generate a latest distribution?

dblock commented:
> > The 3.0 build gets generated even if not all plugins build with it, so looks like the update SHA doesn't have all the plugins? Try another more recent SHA until the tests all pass, possibly hunt down why the build is failing for the plugins that are failing (geospatial looks like).
> 
> Got it, I had sorted the distributions via the newest and used the latest one already. Can this staging build be triggered manually to generate a latest distribution?

You should try with a newer SHA. The builds happen every day.

If the build is broken (other tests failing, please try another SHA or chase that down).

dblock commented:
Looks like the style checker is failing for good reason (use of indices instead of indexes, etc.).

nhtruong commented:
> Looks like the style checker is failing for good reason (use of indices instead of indexes, etc.).

We have a namespace called `indices` that has been around since the beginning of time. Shouldn't it be the other way around where the use indices over indexes?

dblock commented:
> > Looks like the style checker is failing for good reason (use of indices instead of indexes, etc.).
> 
> We have a namespace called `indices` that has been around since the beginning of time. Shouldn't it be the other way around where the use indices over indexes?

The descriptive text was standardized as "indexes", but technically it's "indices" in APIs and such I believe. Let's open an issue in docs about potentially changing that, but for now the style checker is correct.

gargharsh3134 commented:
Test run against 3.0.0 build is the only failing task. Newly added spec tests though are passing, the existing plugin related tests are failing.

gargharsh3134 commented:
@dblock @nhtruong Can you please help with suggestions on 3.0.0 build? I have picked up the latest SHA as part of this PR.
Thanks!

dblock commented:
Looks like we don't have a good 3.0 build since with all the plugins, hence the failures :(

Undo the 3.0 hash change and add `version: '< 3.0'` in the new list tests, maybe with a `# TODO: re-enable in 3.0, see  https://github.com/opensearch-project/opensearch-api-specification/pull/613`.

gargharsh3134 commented:
> Undo the 3.0 hash change and add version: '< 3.0' in the new list tests, maybe with a # TODO: re-enable in 3.0, see  https://github.com/opensearch-project/opensearch-api-specification/pull/613.

@dblock Thanks for the suggestions. The tasks are green now, please take a look whenever you get a chance.
Thanks!

dblock commented:
Good work @gargharsh3134, thanks for hanging in here!
